near-sdk/src/store/free_list/mod.rs

```diff
@@ -220,6 +220,84 @@ where
     pub fn drain(&mut self) -> Drain<T> {
         Drain::new(self)
     }
+
+    /// Empty slots in the front of the list is swapped with occupied slots in back of the list.
+    /// Defrag helps reduce gas cost in certain scenarios where lot of elements in front of the list are
+    /// removed without getting replaced. Please see https://github.com/near/near-sdk-rs/issues/990
+    pub fn defrag(&mut self) -> HashMap<u32, u32> {
+        let elements_size = self.elements.len();
+        let mut forward_ix = 0;
+        let mut backward_ix = elements_size - 1;
+        let mut swaps: HashMap<u32, u32> = HashMap::default();
+        while forward_ix < backward_ix {
+            //increment forward index until empty slot is found
```

Comment:
You can traverse this chain using `self.first_free` and then pulling the next indices from each element loaded. This _should_ contain all of the freed elements (given no bugs anywhere). This avoids the need to traverse everything manually

Reply:
I don't understand what you are explaining. 

1. `self.first_free` would be last removed. If items are being removed at the start of iterator then the `n`th element removed (at index `n-1`) would be self.first_free. It might not be efficient to start replacing there. I thought we want to populate from the front.
2. Should we design this for a generic scenario of fragmentation or only for the scenario described in the issue?

Reply:
> I thought we want to populate from the front.

Nope, the freelist is unordered; it doesn't matter where the values are shifted from and to in the underlying `Vector`

> It might not be efficient to start replacing there

The assumption here, I assume, is that you won't have to move values twice if the array shrinks, but this all would happen in-memory, so it would be cheap. What you could do is traverse all freed nodes to get the indices and check for this case so that you only move the values at most once. This avoids loading values that will not be moved with this operation.

> Should we design this for a generic scenario of fragmentation or only for the scenario described in the issue?

I don't know what you mean by this; can you clarify?

Reply:
For a scenario like below, we will end up swapping Occupied1 and Occupied2 with Empty1 and Empty2 with first two entries (Empty4 and Empty3) still empty. For efficient defragmentation we might have to start from the beginning. Did I understand your approach correctly?

Empty4, Empty3, Empty2, Empty1=(self.first_free), Occupied1, Occupied2

Reply:
Let me update the commit with checking `< self.occupied_count`.